### PR TITLE
Python Bindings Tweaks and New Failing Test

### PR DESCRIPTION
**Pull request by Eric-Vin**
Added the ability to pass hash and cell count to sampler and fixed minor parsing issue for Python keyword arguments in setup_sampler.

I also added a test to check validity of generated samples. It generates a series of samples and checks that the formula is still solvable when variables are fixed to the sampled values. It currently fails on one of the generated samples. Please let me know if this test doesn't make sense, since that could definitely be the case. I removed the old tests since they were for the previous version of the Python bindings before you updated them.

### CODE REVIEW

File: python/src/pyunigen.cpp [hunk 1]
```diff
@@ -139,7 +139,7 @@ static void setup_sampler(Sampler *self, PyObject *args, PyObject *kwds)
     self->kappa = self->unig->get_kappa();
 
     static char* kwlist[] = {"verbosity", "seed", "epsilon", "delta", "kappa", "multisample", NULL};
-    if (!PyArg_ParseTupleAndKeywords(args, kwds, "|iIddp", kwlist,
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "|iIdddp", kwlist,
         &self->verbosity, &self->seed, &self->epsilon, &self->delta, &self->kappa, &self->multisample))
```
Comment (msoos) [inline python/src/pyunigen.cpp:143]: Good catch!!

File: python/src/pyunigen.cpp [hunk 2]
```diff
@@ -321,7 +328,15 @@ static PyObject* sample(Sampler *self, PyObject *args, PyObject *kwds)
     }
 
     self->appmc->set_projection_set(self->sampling_set);
-    auto sol_count = self->appmc->count();
+    
+    auto sol_count = * new ApproxMC::SolCount;
```
Comment (msoos) [inline python/src/pyunigen.cpp:332]: Don't create a `new` here. Just do
```
ApproxMC::SolCount solcount;
```

Then overwrite it if need be. You never free sol_count, for example. No need for dynamic allocation here.

**Comment by msoos** [comment]
Thanks! What did you mean by ` I was looking through the code of the pyunigen bindings trying to find the cause, and I noticed that when importing the variables in the Python bindings you subtract 1 from them first` -- where is this subtraction? I can't find it...

File: python/src/pyunigen.cpp [hunk 3]
```diff
@@ -306,7 +310,10 @@ static PyObject* sample(Sampler *self, PyObject *args, PyObject *kwds)
     }
 
     PyObject* sampling_set_obj = NULL;
-    if (!PyArg_ParseTupleAndKeywords(args, kwds, "|IO", kwlist, &self->samples_needed, &sampling_set_obj))
+    int hash_count = NULL;
+    int cell_count = NULL;
```
Comment (msoos) [inline python/src/pyunigen.cpp:314]: You should not initialize integers with NULL. NULL is for pointers, it's misleading.

**Comment by Eric-Vin** [comment]
> Thanks! What did you mean by ` I was looking through the code of the pyunigen bindings trying to find the cause, and I noticed that when importing the variables in the Python bindings you subtract 1 from them first` -- where is this subtraction? I can't find it...

Ah that's a minor thing I'm noticing like on line 118 with `long var = val - 1;` I don't think that's causing the issue though, since everything is being returned correctly.

File: python/src/pyunigen.cpp [hunk 4]
```diff
@@ -321,7 +328,15 @@ static PyObject* sample(Sampler *self, PyObject *args, PyObject *kwds)
     }
 
     self->appmc->set_projection_set(self->sampling_set);
-    auto sol_count = self->appmc->count();
+    
+    auto sol_count = * new ApproxMC::SolCount;
+    if (hash_count != NULL && cell_count != NULL) {
```
Comment (msoos) [inline python/src/pyunigen.cpp:333]: This is incorrect check. If the user sets them to 0 & 0, the system will assume the user didn't set them. That's not what's needed here. What you actually wanted to check is whether the option was passed at all. It should ask for a single Python Object (a tuple of 2 elements), and if it's NULL then the object was not passed by the user. If it is not NULL then we know it's passed, we unpack and we use it.

Comment (msoos) [reply python/src/pyunigen.cpp:333]: Notice that the user could only set 1 of the 2 otherwise, which is nonsensical :)

**Comment by msoos** [comment]
> > Thanks! What did you mean by ` I was looking through the code of the pyunigen bindings trying to find the cause, and I noticed that when importing the variables in the Python bindings you subtract 1 from them first` -- where is this subtraction? I can't find it...
> 
> Ah that's a minor thing I'm noticing like on line 118 with `long var = val - 1;` I don't think that's causing the issue though, since everything is being returned correctly.

Haha, that was written by you, and it's correct actually! I cleaned up the function, though. NULL return always means ERROR in Python. So I reversed the way it works.

**Comment by msoos** [comment]
Can you please fix the issues I pointed to above? I will then merge and fix.

Thanks,

Mate

**Comment by msoos** [comment]
(BTW, I fixed the build of the Python package now, so once I merge it should build automatically, yay! See: https://github.com/meelgroup/unigen/actions/runs/3424213367 )

**Comment by msoos** [comment]
I fixed it all up. I tried giving feedback above, but I also need to get this off my table. It's merged and pushed, and the relevant commits are under your name.

Mate

**Pull request merged**